File: tests/flare_emu_test_hooks.c
/*
Copyright (C) 2018 FireEye, Inc.

Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-BSD-3-CLAUSE or
https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be
copied, modified, or distributed except according to those terms.

Author: James T. Bennett

unit tests for flare-emu
tests for iterate and runtime/Windows API function hook features of flare-emu
*/

#include <stdio.h>
#include <Windows.h>

int main()
{
	char str[] = "this is a test";
	wchar_t wstr[50];
	char str2[50];
	char *str3, *str4;
	wchar_t *wstr2, *wstr3;
	int ret;
	BOOL flag;
	HANDLE hHandle;
	ret = MultiByteToWideChar(CP_ACP, 0, str, -1, wstr, 50);
	printf("%S is %d bytes from MultiByteToWideChar\r\n", wstr, ret);
	ret = WideCharToMultiByte(CP_ACP, 0, wstr, -1, str2, sizeof(str2), NULL, &flag);
	printf("%s is %d bytes from WideCharToMultiByte\r\n", str2, ret);
	ZeroMemory(wstr, sizeof(wstr));
	ZeroMemory(str2, sizeof(str2));
	ret = MultiByteToWideChar(CP_ACP, 0, str, 4, wstr, 50);
	printf("%S is %d bytes from truncated MultiByteToWideChar\r\n", wstr, ret);
	ret = WideCharToMultiByte(CP_ACP, 0, wstr, 4, str2, sizeof(str2), NULL, &flag);
	printf("%s is %d bytes from truncated WideCharToMultiByte\r\n", str2, ret);
	hHandle = GetProcessHeap();
	if (!hHandle)
		return -1;
	str3 = (char*)HeapAlloc(hHandle, NULL, 0x1000);
	strcpy(str3, str);
	printf("%s strcpy to HeapAlloc\r\n", str3);
	ZeroMemory(str3, 0x1000);
	lstrcpy(str3, str);
	printf("%s lstrcpy to HeapAlloc\r\n", str3);
	str3 = (char*)HeapReAlloc(hHandle, NULL, str3, 0x2000);
	printf("%s from HeapReAlloc\r\n", str3);
	HeapFree(hHandle, NULL, str3);
	str3 = (char*)LocalAlloc(LMEM_FIXED, 0x1000);
	strncpy(str3, str, sizeof(str));
	printf("%s strncpy to fixed LocalAlloc\r\n", str3);
	strncpy(str3, str2, sizeof(str2));
	printf("%s strncpy to fixed LocalAlloc with padding\r\n", str3);
	LocalFree(str3);
	hHandle = LocalAlloc(LMEM_MOVEABLE, 0x1000);
	str3 = (char*)LocalLock(hHandle);
	strncpy_s(str3, 0x1000, str, sizeof(str));
	printf("%s strncpy_s to movable LocalAlloc\r\n", str3);
	LocalUnlock(hHandle);
	hHandle = LocalReAlloc(hHandle, 0x2000, NULL);
	str3 = (char*)LocalLock(hHandle);
	printf("%s from LocalReAlloc\r\n", str3);
	LocalFree(hHandle);
	VirtualAlloc((LPVOID)0x99000, 0x2000, MEM_RESERVE, PAGE_READWRITE);
	wstr2 = (wchar_t*)VirtualAlloc((LPVOID)0x99000, 0x2000, MEM_COMMIT, PAGE_READWRITE);
	ret = mbstowcs(wstr2, str, sizeof(str));
	printf("%S mbstowcs (returned %d) to double VirtualAlloc\r\n", wstr2, ret);
	VirtualFree((LPVOID)wstr2, 0x2000, MEM_RELEASE);
	wstr2 = (wchar_t*)VirtualAlloc(NULL, 0x2000, MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE);
	ret = mbtowc(wstr2, str, sizeof(str));
	printf("%S mbtowc (returned %d) to single VirtualAlloc\r\n", wstr2, ret);
	VirtualFree((LPVOID)wstr2, 0x2000, MEM_RELEASE);
	hHandle = GetCurrentProcess();
	wstr2 = (wchar_t*)VirtualAllocEx(hHandle, NULL, 0x1000, MEM_RESERVE | MEM_COMMIT, PAGE_READWRITE);
	wcscpy(wstr2, wstr);
	printf("%S wcscpy to VirtualAllocEx\r\n", wstr2);
	VirtualFreeEx(hHandle, wstr2, 0x1000, MEM_RELEASE);
	wstr2 = (wchar_t*)malloc(0x1000);
	wcsncpy(wstr2, wstr, sizeof(wstr));
	printf("%S wcsncpy to malloc\r\n", wstr2);
	mbstowcs(wstr2, str, sizeof(str));
	wcsncpy(wstr2, L"test", 20);
	printf("%S wcsncpy to malloc with padding\r\n", wstr2);
	free(wstr2);
	wstr2 = (wchar_t*)calloc(0x1000, 2);
	wcsncpy_s(wstr2, 0x1000, wstr, sizeof(wstr));
	printf("%S wcsncpy_s to calloc\r\n", wstr2);
	memcpy(wstr2 + 20, str + 10, 5);
	printf("%s memcpy to offset\r\n", (char*)(wstr2 + 20));
	printf("%s is %d characters in strlen\r\n", str, strlen(str));
	printf("%s is not really %d characters in strnlen\r\n", str, strnlen(str, 2));
	wcsncpy(wstr2, wstr, sizeof(wstr));
	printf("%S is %d characters in wcslen\r\n", wstr2, wcslen(wstr2));
	printf("%S is not really %d characters in wcsnlen\r\n", wstr2, wcsnlen(wstr2, 2));
	free(wstr2);
	str3 = (char*)malloc(0x1000);
	strcpy(str3, str);
	printf("%s strcmp %s result is %d\r\n", str3, str, strcmp(str3, str));
	_strupr(str3);
	printf("%s stricmp %s result is %d\r\n", str3, str, _stricmp(str3, str));
	strcpy(str3, str);
	strcpy(str3 + 10, "mess");
	printf("%s strncmp %s 10 chars result is %d\r\n", str3, str, strncmp(str3, str, 10));
	_strupr(str3);
	printf("%s strnicmp %s 10 chars result is %d\r\n", str3, str, _strnicmp(str3, str, 10));
	free(str3);
	wstr2 = (wchar_t*)malloc(0x1000);
	mbstowcs(wstr, str, sizeof(str));
	mbstowcs(wstr2, str, sizeof(str));
	printf("%S wcscmp %S result is %d\r\n", wstr2, wstr, wcscmp(wstr2, wstr));
	_wcsupr(wstr2);
	printf("%S wcsicmp %S result is %d\r\n", wstr2, wstr, _wcsicmp(wstr2, wstr));
	wcscpy(wstr2, wstr);
	wcscpy(wstr2 + 10, L"mess");
	printf("%S wcsncmp %S 10 chars result is %d\r\n", wstr2, wstr, wcsncmp(wstr2, wstr, 10));
	_wcsupr(wstr2);
	printf("%S wcsnicmp %S 10 chars result is %d\r\n", wstr2, wstr, _wcsnicmp(wstr2, wstr, 10));
	printf("%c is strchr index %d in %s\r\n", 'a', ((unsigned int)strchr(str, 'a') - (unsigned int)str), str);
	printf("%C is wcschr index %d in %S\r\n", L'a', ((unsigned int)wcschr(wstr, L'a') - (unsigned int)wstr) / 2, wstr);
	printf("%c is strrchr index %d in %s\r\n", 't', ((unsigned int)strrchr(str, 't') - (unsigned int)str), str);
	printf("%C is wcsrchr index %d in %S\r\n", L't', ((unsigned int)wcsrchr(wstr, L't') - (unsigned int)wstr) / 2, wstr);
	free(wstr2);
	str3 = (char*)malloc(0x1000);
	memset(str3, 0, 0x1000);
	strcat(str3, "The Quick Brown Fox ");
	strncat(str3, "Jumps Over ", 100);
	strncat(str3, "The Lazy DogAAAAAAAAA", 12);
	printf("strcat %s\r\n", str3);
	printf("strlwr %s\r\n", _strlwr(str3));
	wstr2 = (wchar_t*)malloc(0x1000);
	memset(wstr2, 0, 0x1000);
	wcscat(wstr2, L"The Quick Brown Fox ");
	wcsncat(wstr2, L"Jumps Over ", 100);
	wcsncat(wstr2, L"The Lazy DogAAAAAAAAA", 12);
	printf("wcscat %S\r\n", wstr2);
	printf("wcslwr %S\r\n", _wcslwr(wstr2));
	str4 = _strdup(str3);
	memset(str3, 0, 0x1000);
	printf("strdup %s\r\n", str4);
	wstr3 = _wcsdup(wstr2);
	memset(wstr2, 0, 0x1000);
	printf("wcsdup %S\r\n", wstr3);
	free(str4);
	free(wstr3);
	free(str3);
	free(wstr2);
	return 0;
}

File: tests/flare_emu_test_hooks.py
############################################
# Copyright (C) 2018 FireEye, Inc.
#
# Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
# http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-BSD-3-CLAUSE or
# https://opensource.org/licenses/BSD-3-Clause>, at your option. This file may not be
# copied, modified, or distributed except according to those terms.
#
# Author: James T. Bennett
#
# flare_emu_test_hooks.py is a script for testing flare-emu with IDA Pro and Radare2
#
# NOTE: you may have to rename functions IDA Pro fails to recognize, such as printf
#
# Dependencies:
# https://github.com/fireeye/flare-emu
############################################

from __future__ import print_function
import flare_emu
import sys
import logging

tests = {"from MultiByteToWideChar\r\n":["this is a test".encode("utf-16"), 15], 
         "from WideCharToMultiByte\r\n":["this is a test", 15], 
         "truncated MultiByteToWideChar":["this".encode("utf-16"), 4], 
         "truncated WideCharToMultiByte":["this", 4], 
         "strcpy to HeapAlloc":["this is a test"], 
         "lstrcpy to HeapAlloc":["this is a test"], 
         "HeapReAlloc":["this is a test"], 
         "fixed LocalAlloc\r\n":["this is a test"], 
         "fixed LocalAlloc with padding":["this"], 
         "movable LocalAlloc":["this is a test"], 
         "LocalReAlloc":["this is a test"], 
         "mbstowcs":["this is a test".encode("utf-16"), 14], 
         "mbtowc":["t".encode("utf-16"), 1], 
         "VirtualAllocEx":["this".encode("utf-16")], 
         "malloc\r\n":["this".encode("utf-16")], 
         "malloc with padding":["test".encode("utf-16")], 
         "calloc":["this".encode("utf-16")], 
         "memcpy to offset":["test"], 
         "strlen":["this is a test", 14], 
         "strnlen":["this is a test", 2], 
         "wcslen":["this".encode("utf-16"), 4], 
         "wcsnlen":["this".encode("utf-16"), 2], 
         "strcmp":["this is a test", "this is a test", 0], 
         "stricmp":["THIS IS A TEST", "this is a test", 0], 
         "strncmp":["this is a mess", "this is a test", 0], 
         "strnicmp":["THIS IS A MESS", "this is a test", 0], 
         "wcscmp":["this is a test".encode("utf-16"), "this is a test".encode("utf-16"), 0], 
         "wcsicmp":["THIS IS A TEST".encode("utf-16"), "this is a test".encode("utf-16"), 0], 
         "wcsncmp":["this is a mess".encode("utf-16"), "this is a test".encode("utf-16"), 0], 
         "wcsnicmp":["THIS IS A MESS".encode("utf-16"), "this is a test".encode("utf-16"), 0], 
         "strchr":[97, 8, "this is a test"], 
         "wcschr":[97, 8, "this is a test".encode("utf-16")], 
         "strrchr":[116, 13, "this is a test"], 
         "wcsrchr":[116, 13, "this is a test".encode("utf-16")], 
         "strcat":["The Quick Brown Fox Jumps Over The Lazy Dog"], 
         "strlwr":["the quick brown fox jumps over the lazy dog"], 
         "wcscat":["The Quick Brown Fox Jumps Over The Lazy Dog".encode("utf-16")], 
         "wcslwr":["the quick brown fox jumps over the lazy dog".encode("utf-16")], 
         "strdup":["the quick brown fox jumps over the lazy dog"], 
         "wcsdup":["the quick brown fox jumps over the lazy dog".encode("utf-16")]
        }
 
def iterateHook(eh, address, argv, userData):
    testString = eh.getEmuString(argv[0]).decode("latin1")
    for test in tests:
        if test in testString:
            print("testing '%s'" % testString.replace("\r\n", ""))
            for i in range(len(tests[test])):
                if isinstance(tests[test][i], bytes) or isinstance(tests[test][i], str) :
                    if tests[test][i][:2] == b"\xff\xfe":
                        expected = tests[test][i][2:]
                        actual = eh.getEmuWideString(argv[i+1])
                    else:
                        expected = tests[test][i]
                        actual = eh.getEmuString(argv[i+1]).decode("latin1")
                else:
                    expected = tests[test][i]
                    actual = argv[i+1]
                
                if expected != actual:
                    print("FAILED: %s does not match expected result %s" % (actual,expected))
            return
    print("%s: test not found" % (testString.replace("\r\n", "")))
    
def wcsdupHook(eh, address, argv, funcName, userData):
    print("the new wcsdup hook was called")
    eh.hookCalled = True
    if eh.isValidEmuPtr(argv[0]):
        s = "the quick brown fox jumps over the lazy dog".encode("utf-16")[2:]
        memAddr = eh.allocEmuMem(len(s) + 2)
        eh.uc.mem_write(memAddr, s)
        eh.uc.reg_write(eh.regs["ret"], memAddr)
        return
    
    eh.uc.reg_write(eh.regs["ret"], 0)

if __name__ == '__main__':     
    # optional argument with sample path to test radare2 support
    if len(sys.argv) == 2:
        eh = flare_emu.EmuHelper(samplePath=sys.argv[1])
    else:
        eh = flare_emu.EmuHelper()
        
    eh.analysisHelper.setName(0x41141a, "printf")
    print("testing iterate feature for printf function")
    strcpyEa = eh.analysisHelper.getNameAddr("j_strcpy")
    eh.iterate(eh.analysisHelper.getNameAddr("printf"), iterateHook)
    eh.analysisHelper.setName(strcpyEa, "testname")
    eh.addApiHook("testname", "strcpy")
    eh.addApiHook("wcsdup", wcsdupHook)
    print("testing with renamed and redirected strcpy hook and new hook for wcsdup")
    eh.iterate(eh.analysisHelper.getNameAddr("printf"), iterateHook)
    if "hookCalled" not in dir(eh):
        print("FAILED: addApiHook hook not called")
    
